File: kalitech-engine/src/main/resources/kalitech/engine/modules/Controllers/EntityControllerLink.js
// FILE: @module/Controllers/EntityControllerLink.js
"use strict";

function req(v, msg) {
    if (v == null) throw new Error(msg);
    return v;
}

class EntityControllerLink {
    constructor(id, entity, controller) {
        this.id = String(id || "");
        this.entity = req(entity, "[EntityControllerLink] entity is required");
        this.controller = req(controller, "[EntityControllerLink] controller is required");
        this._alive = true;
    }

    setController(controller) {
        if (!this._alive) throw new Error("[EntityControllerLink] setController on disposed");
        this.controller = req(controller, "[EntityControllerLink] controller is required");
        return this;
    }

    update(dt) {
        if (!this._alive) throw new Error("[EntityControllerLink] update on disposed");
        const c = this.controller;
        if (c && typeof c.update === "function") c.update(dt);
    }

    dispose() {
        if (!this._alive) return;
        this._alive = false;

        const c = this.controller;
        this.controller = null;

        if (c && typeof c.dispose === "function") c.dispose();

        this.entity = null;
        this.id = "";
    }
}

module.exports = {EntityControllerLink};

File: kalitech-engine/src/main/resources/kalitech/engine/modules/Controllers/helpers/EntityControllerLink.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local Classes = luaRuntime.class
local Error = luaRuntime.Error
function req(self, v, msg)
    if v == nil then
        error(
            Classes:construct(Error, msg),
            0
        )
    end
    return v
end
EntityControllerLink = Classes:create()
EntityControllerLink.name = "EntityControllerLink"
function EntityControllerLink.prototype.lua_constructor(self, id, entity, controller)
    self.id = tostring(id or "")
    self.entity = req(_G, entity, "[EntityControllerLink] entity is required")
    self.controller = req(_G, controller, "[EntityControllerLink] controller is required")
    self._alive = true
end
function EntityControllerLink.prototype.setController(self, controller)
    if not self._alive then
        error(
            Classes:construct(Error, "[EntityControllerLink] setController on disposed"),
            0
        )
    end
    self.controller = req(_G, controller, "[EntityControllerLink] controller is required")
    return self
end
function EntityControllerLink.prototype.update(self, dt)
    if not self._alive then
        error(
            Classes:construct(Error, "[EntityControllerLink] update on disposed"),
            0
        )
    end
    local c = self.controller
    if c and KTypeOf(c.update) == "function" then
        c:update(dt)
    end
end
function EntityControllerLink.prototype.dispose(self)
    if not self._alive then
        return
    end
    self._alive = false
    local c = self.controller
    self.controller = nil
    if c and KTypeOf(c.dispose) == "function" then
        c:dispose()
    end
    self.entity = nil
    self.id = ""
end
M = {EntityControllerLink = EntityControllerLink}

return M
